src/cryptonote_core/blockchain.cpp

```diff
@@ -576,6 +576,26 @@ bool Blockchain::deinit()
   return true;
 }
 //------------------------------------------------------------------
+// This function removes blocks from the top of blockchain.
+// It starts a batch and calls private method pop_block_from_blockchain().
+void Blockchain::pop_blocks(uint64_t nblocks)
+{
+  CRITICAL_REGION_LOCAL(m_tx_pool);
+  CRITICAL_REGION_LOCAL1(m_blockchain_lock);
+
+  while (!m_db->batch_start())
+  {
+    epee::misc_utils::sleep_no_w(1000);
```

Comment:
Unlock/lock here as already done in blockchain.cpp when starting a batch. I'm not sure exactly why this is needed atm but I remember knowing at some point :)

Reply:
It is not a good idea to sleep when you are holding 2 locks ...  
By unlocking before sleeping, it gives chances to other threads to hold the locks and do their jobs.